matrix_sdk_base/src/models/room.rs

```diff
@@ -631,28 +632,79 @@ impl Room {
         {
-            msg.content = AnyMessageEventContent::RoomRedaction(event.content.clone());
-
-            let redaction =
-                redaction_event_from_redaction_stub(event.clone(), self.room_id.clone());
-            msg.unsigned.redacted_because = Some(EventJson::from(redaction));
+            match msg.deref_mut() {
+                FullOrRedactedEvent::Full(event) => match event {
+                    AnyMessageEventStub::RoomMessage(event) => {
+                        let MessageEventStub {
+                            content,
+                            event_id,
+                            sender,
+                            origin_server_ts,
+                            mut unsigned,
+                        } = event.clone();
+                        // TODO can redacted events that are stripped or have no room_id be
+                        // added to the `redacted_because` field?
+                        unsigned.redacted_because =
+                            Some(EventJson::from(redaction_event_from_redaction_stub(
+                                redacted_event.clone(),
+                                self.room_id.clone(),
+                            )));
+                        let redacted = content.redact();
+                        msg.0 = FullOrRedactedEvent::Redacted(
+                            AnyRedactedMessageEventStub::RoomMessage(RedactedMessageEventStub {
+                                content: redacted,
+                                event_id,
+                                origin_server_ts,
+                                sender,
+                                unsigned,
+                            }),
+                        )
+                    }
+                    AnyMessageEventStub::Sticker(event) => {
+                        let MessageEventStub {
+                            content,
+                            event_id,
+                            sender,
+                            origin_server_ts,
+                            mut unsigned,
+                        } = event.clone();
+                        unsigned.redacted_because =
+                            Some(EventJson::from(redaction_event_from_redaction_stub(
+                                redacted_event.clone(),
+                                self.room_id.clone(),
+                            )));
+                        let redacted = content.redact();
+                        msg.0 = FullOrRedactedEvent::Redacted(AnyRedactedMessageEventStub::Sticker(
+                            RedactedMessageEventStub {
+                                content: redacted,
+                                event_id,
+                                origin_server_ts,
+                                sender,
+                                unsigned,
+                            },
+                        ))
+                    }
+                    // TODO handle the rest of the message events
```

Comment:
This will end up having a branch for every event type? Isn't the server side interested in such a feature as well? Perhaps we'll want this in Ruma if they are.

Reply:
Ruma handle this as in having the event enums have a method like
```rust
fn redact(&mut self, redaction_event: RedactionEvent) -> Self {
    // add to the unsigned field and redact the content
}
```
I think we may have talked about adding something like this. @jplatte would this make sense in ruma?

Reply:
What is "this" exactly? I can tell that this code does redactions and returns whether it was a no-op or not, but I can't tell what `FullOrRedactedEvent` is, is it similar to `AnyRoomEvent`?

Reply:
FullOrRedactedEvent is
```rust
pub enum FullOrRedactedEvent {
    Full(AnyMessageEventStub),
    Redacted(AnyRedactedMessageEventStub),
}
```
which would be obsolete after https://github.com/ruma/ruma/issues/119

Reply:
Okay so I think something that we could definitely have is

```rust
impl AnyMessageEvent {
    pub fn redact(self, redaction: RedactionEvent) -> AnyRedactedMessageEvent { … }
}

impl AnyMessageEventStub {
    pub fn redact(self, redaction: RedactionEventStub) -> AnyRedactedMessageEventStub { … }
}

impl AnyStateEvent {
    pub fn redact(self, redaction: RedactionEvent) -> AnyRedactedStateEvent { … }
}

impl AnyStateEventStub {
    pub fn redact(self, redaction: RedactionEventStub) -> AnyRedactedStateEventStub { … }
}

impl AnyStrippedStateEventStub {
    pub fn redact(self, redaction: RedactionEventStub) -> AnyRedactedStrippedStateEventStub { … }
}
```

which should eliminate most of the boilerplate here, right?

Reply:
Yup, that would be perfect! That would remove the `match` completely.